lib/memo_wise.rb

```diff
@@ -192,33 +200,36 @@ def #{method_name}
 
             if has_arg && MemoWise::InternalAPI.has_kwarg?(method)
               args_str = "(*args, **kwargs)"
-              fetch_key = "[args, kwargs].freeze"
+              fetch_key_init = "key = [args, kwargs].hash"
             elsif has_arg
               args_str = "(*args)"
-              fetch_key = "args"
+              fetch_key_init = "key = args.hash"
             else
               args_str = "(**kwargs)"
-              fetch_key = "kwargs"
+              fetch_key_init = "key = kwargs.hash"
             end
+            fetch_key = "key"
           end
 
-          # Note that we don't need to freeze args before using it as a hash key
-          # because Ruby always copies argument arrays when splatted.
           klass.module_eval <<-END_OF_METHOD, __FILE__, __LINE__ + 1
             # def foo(*args, **kwargs)
-            #   hash = @_memo_wise.fetch(:foo) do
-            #     @_memo_wise[:foo] = {}
-            #   end
-            #   hash.fetch([args, kwargs].freeze) do
-            #     hash[[args, kwargs].freeze] = _memo_wise_original_foo(*args, **kwargs)
+            #   hash = @_memo_wise[:foo] ||= {}
+            #   key = [args, kwargs].hash
```

Comment:
To be clear: this does the following:

1. Create the `Array [args, kwargs]`
2. Calls method `Array#hash` on it, which delegates to both args.hash and kwargs.hash internally
3. Throws away the array from (1) as soon as GC runs, no reference is held to it
4. Uses the hash value as the key

IF this is accruate: is there any advantage to trying to avoid the allocation of the 2-element array in 1?

Reply:
I had the same thought process @ms-ati. I don't know enough about hash functions to feel that confident implementing our own here (i.e. something like `args.hash ^ kwargs.hash`), though I'm open to it. I did look at the [`Array#hash` source](https://ruby-doc.org/core-3.0.1/Array.html#method-i-hash) but was wary of trying to implement something like that in Ruby and missing out on future improvements to the built-in `Array#hash`. Thoughts?

Reply:
Update: using XOR is indeed a noticeable speed boost. @ms-ati @jemmaissroff do we think this is safe to use or should we worry about hash collisions?

Reply:
Update: I no longer think we should use XOR, because I just realized it does not handle ordering, so `foo(1, 2)` and `foo(2, 1)` would collide. Still open to other options that avoid the array allocation though!

Reply:
Based on https://crypto.stackexchange.com/a/55219, @jemmaissroff and I played around with doing things like

```ruby
(args.hash << 64) & kwargs.hash).hash
```

which is indeed faster.

I don't know enough about:
- **hashing** to say whether this is safe from a collision perspective
- **ruby** to say whether 64 is something platform-dependent here

Reply:
That is definitely not right. It should be `|` instead of `&`. The idea seems to be that we assume the hash function returns a 64 bit number. Shift it by 64 bits to the left to make room for the second number. Then we need to use bitwise `or`, not `and` to put the two numbers together. Then get the has of the result to get back into the range of 64 bits.

I'm certainly not an expert on this, so I can't say if it's a legitimate algorithm or not. But I do know about the [Cantor pairing function](https://en.wikipedia.org/wiki/Pairing_function). It might be worth checking out.

Reply:
🤦‍♀️ yes I meant to write `|`—thank you for the correction! I'll look into the Cantor pairing function—thanks for the link! @ms-ati I'm curious if you have any knowledge here.

Reply:
Hi @JacobEvelyn -- Hmm stepping back, I know the goal is performance to avoid the ephemeral allocation of the `Array`. But using only a `#hash` as the memoization key does invite collision, and investigating the collisions.

Are there any alternatives that **also avoid the allocation** but retain the full ability to avoid collision of memoized inputs?

Reply:
Update: I don't think we should use

```ruby
(args.hash << 64) | kwargs.hash).hash
```

as I'm seeing lots of hash collisions with that approach.

Reply:
Going to table this for now and potentially explore it again in #190